README.md
# Python inara.cz interface for Elite Dangerous.

(c). Ray C Horn, All RIghts Reserved.



django-standalone-0.4/CHANGES
version 0.4: (released Mar 7th, 2010)
	- testing integrated with setup.py
	- only script models are now patched into standalone.model
	- fixed a bug with installed apps not working right if the script
	  prepopulated INSTALLED_APPS

version 0.3: (released Feb 28th, 2010)
	- access to the django REPL with access to models
	- added very simple testing to the project
	- added a way to do sdist

version 0.2 fixed setup.py

version 0.1 first public version with missing dependency in setup.py


main.py
import os
import sys
import pickle
import json
#print("PYTHONPATH = {}".format(os.environ['PYTHONPATH']))

paths = os.environ['PYTHONPATH'].split(os.pathsep)
for p in paths:
    if (p not in sys.path):
        print("Adding {} to sys.path.".format(p))
        sys.path.insert(0, p)

from vyperlogix import _utils

if (0):
    for f in sys.path:
        print(f)

from inara import sample as inara_sample
from inara import scrape_commodities
from inara import scrape_commodity_data
from inara import upload_to_google_drive

__scrape_commodities__ = False
__scrape_commodity_data__ = True
__prep_google_creds__ = False
__upload_to_google__ = False

if (__name__ == '__main__'):
    # float(sys.version_info.major)+(float(sys.version_info.minor)/10)+(float(sys.version_info.micro)/100)
    py_version = _utils.getVersionFloat()
    if (py_version < 3.8) or (py_version > 3.9):
        print('ERROR: Requires Python 3.8.x rather than {}. Please use the correct Python version.'.format(
            py_version))
        sys.exit(-1)

    if (__scrape_commodities__):
        scrape_commodities()
        from commodities import Commodities
        from commodities import all_commodities
        from commodities import commodities
    if (__scrape_commodity_data__):
        from queue import Queue
        import commodities

        target_dirname = './data'
        if (0):
            import time
            from concurrent import futures as concurrent_futures
            
            __done__ = {'is_done':False}
            def wait_for_output(output_queue, __done__):
                while (1):
                    try:
                        msg = output_queue.get(timeout=5)
                        if (msg):
                            sys.stdout('{}\n'.format(msg))
                        if (__done__.get('is_done', False) and (output_queue.qsize() == 0)):
                            sys.stderr.write('Nothing more to do.\n')
                            break
                    except:
                        __done__['is_done'] = True
                    time.sleep(1)
            
            output = Queue(maxsize = 100)
            items = [wait_for_output, commodities.commodities_by_name.get('Tritium'), commodities.commodities_by_name.get('AgronomicTreatment')]
    
            with concurrent_futures.ThreadPoolExecutor(max_workers=10) as executor:
                futures = {executor.submit(scrape_commodity_data, commodity_refid=item, star_system_refid=0, dirname=target_dirname, is_verbose=False, fOut=output) if (not isinstance(item, callable)) else item(output, __done__): item for item in items}
                for future in concurrent_futures.as_completed(futures):
                    try:
                        data = future.result()
                    except Exception as ex:
                        print('Got an exception: %s' % (ex))
                sys.stderr.write('Signal we are done.\n')
                __done__['is_done'] = True
        else:
            items = [commodities.commodities_by_name.get('Tritium'), commodities.commodities_by_name.get('AgronomicTreatment')]
            for item in items:
                __commodity_name__ = commodities.commodities_by_value.get(item, 'UNKNWON-COMMODITY')
                fname = 'commodity_{}_report.txt'.format(__commodity_name__)
                fpath = os.sep.join([target_dirname, fname])
                with open(fpath, 'w') as ffOut:
                    scrape_commodity_data(commodity_refid=item, star_system_refid=0, dirname=target_dirname, is_verbose=True, fOut=ffOut)
                    ffOut.flush()

    if (__prep_google_creds__):
        fpath1 = os.path.abspath('./token.pickle')
        fpath2 = os.path.abspath('./client_secrets.json')
        has_file1 = (os.path.exists(fpath1)) and (os.path.isfile(fpath1))
        has_file2 = (os.path.exists(fpath2)) and (os.path.isfile(fpath2))
        if (has_file1):
            if (not has_file2):
                with open(fpath1, 'rb') as token:
                    creds = pickle.load(token)
                    my_creds = json.loads(creds.to_json(strip=None))
                    my_creds["redirect_uris"] = ["urn:ietf:wg:oauth:2.0:oob", "http://localhost"]
                    my_creds["auth_uri"] = "https://accounts.google.com/o/oauth2/auth"
                    my_creds = {"installed": my_creds}
                    with open(fpath2, 'w') as secrets:
                        secrets.write(json.dumps(my_creds))
        else:
            print("ERROR: Missing the creds in {}.".format(fpath1))
            sys.exit(-1)
                
    if (__upload_to_google__):
        fpath = os.path.abspath('./data')
        files = [os.sep.join([fpath, f]) for f in os.listdir(fpath)]
        for f in files:
            upload_to_google_drive(f)


quickstart.py
from __future__ import print_function
import pickle
import os.path
from googleapiclient.discovery import build
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request

# If modifying these scopes, delete the file token.pickle.
SCOPES = ['https://www.googleapis.com/auth/drive.metadata.readonly']

def main():
    """Shows basic usage of the Drive v3 API.
    Prints the names and ids of the first 10 files the user has access to.
    """
    creds = None
    # The file token.pickle stores the user's access and refresh tokens, and is
    # created automatically when the authorization flow completes for the first
    # time.
    if os.path.exists('token.pickle'):
        with open('token.pickle', 'rb') as token:
            creds = pickle.load(token)
    # If there are no (valid) credentials available, let the user log in.
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            creds.refresh(Request())
        else:
            flow = InstalledAppFlow.from_client_secrets_file(
                'credentials.json', SCOPES)
            creds = flow.run_local_server(port=0)
        # Save the credentials for the next run
        with open('token.pickle', 'wb') as token:
            pickle.dump(creds, token)

    service = build('drive', 'v3', credentials=creds)

    # Call the Drive v3 API
    results = service.files().list(
        pageSize=10, fields="nextPageToken, files(id, name)").execute()
    items = results.get('files', [])

    if not items:
        print('No files found.')
    else:
        print('Files:')
        for item in items:
            print(u'{0} ({1})'.format(item['name'], item['id']))

if __name__ == '__main__':
    main()
    

libs/commodities.py
from vyperlogix import enum
class Commodities(enum.Enum.Enum):
    AdvancedCatalysers=61
    AdvancedMedicines=166
    AgriMedicines=1
    AgronomicTreatment=10268
    AIRelics=89
    Alexandrite=10249
    Algae=15
    Aluminium=37
    AncientArtefact=121
    AncientCasket=10153
    AncientKey=10240
    AncientOrb=10154
    AncientRelic=10155
    AncientTablet=10156
    AncientTotem=10157
    AncientUrn=10158
    AnimalMeat=16
    AnimalMonitors=62
    AntimatterContainmentUnit=10167
    AntiqueJewellery=10209
    Antiquities=91
    AquaponicSystems=63
    ArticulationMotors=182
    AssaultPlans=169
    AtmosphericProcessors=87
    AutoFabricators=65
    BasicMedicines=33
    BattleWeapons=88
    Bauxite=51
    Beer=10
    Benitoite=10247
    Bertrandite=52
    Beryllium=38
    BioreducingLichen=66
    Biowaste=76
    Bismuth=106
    BlackBox=122
    BootlegLiquor=95
    Bromellite=148
    BuildingFabricators=102
    CeramicComposites=100
    ChemicalWaste=32
    Clothing=7
    CMMComposite=140
    Cobalt=39
    Coffee=17
    Coltan=55
    CombatStabilisers=34
    CommercialSamples=170
    ComputerComponents=67
    ConductiveFabrics=165
    ConsumerTechnology=8
    Copper=40
    CropHarvesters=29
    Cryolite=110
    DamagedEscapePod=10215
    DataCore=10166
    DiplomaticBag=171
    DomesticAppliances=9
    EarthRelics=10210
    EmergencyPowerCells=158
    EncryptedCorrespondence=172
    EncryptedDataStorage=173
    EnergyGridAssembly=149
    EvacuationShelter=99
    ExhaustManifold=159
    ExperimentalChemicals=123
    Explosives=3
    Fish=18
    FoodCartridges=19
    FossilRemnants=10221
    FruitandVegetables=20
    Gallite=56
    Gallium=41
    GeneBank=10211
    GeologicalEquipment=103
    GeologicalSamples=174
    Gold=42
    Goslarite=111
    Grain=21
    Grandidierite=10248
    HESuits=68
    Hafnium178=124
    HardwareDiagnosticSensor=155
    HeatsinkInterlink=151
    HNShockMount=150
    Hostage=175
    HydrogenFuel=4
    HydrogenPeroxide=138
    ImperialSlaves=49
    Indite=57
    Indium=43
    InsulatingMembrane=141
    IonDistributor=160
    Jadeite=168
    LandEnrichmentSystems=71
    Landmines=118
    Lanthanum=107
    LargeSurveyDataCache=125
    Leather=73
    Lepidolite=58
    LiquidOxygen=137
    Liquor=11
    Lithium=44
    LithiumHydroxide=147
    LowTemperatureDiamonds=144
    MagneticEmitterCoil=152
    MarineEquipment=86
    MedicalDiagnosticEquipment=154
    MetaAlloys=101
    MethaneClathrate=145
    MethanolMonohydrateCrystals=146
    MicroControllers=156
    MicroWeaveCoolingHoses=185
    MicrobialFurnaces=85
    MilitaryGradeFabrics=157
    MilitaryIntelligence=126
    MilitaryPlans=127
    MineralExtractors=31
    MineralOil=5
    ModularTerminals=181
    Moissanite=116
    MolluscBrainTissue=10256
    MolluscFluid=10255
    MolluscMembrane=10252
    MolluscMycelium=10253
    MolluscSoftTissue=10254
    MolluscSpores=10251
    Monazite=10245
    MuonImager=119
    Musgravite=10246
    MysteriousIdol=10219
    Nanobreakers=167
    Narcotics=12
    NaturalFabrics=74
    NeofabricInsulation=183
    NerveAgents=96
    NonLethalWeapons=78
    OccupiedEscapePod=129
    Osmium=72
    Painite=84
    Palladium=45
    PerformanceEnhancers=35
    PersonalEffects=10159
    PersonalWeapons=79
    Pesticides=6
    Platinum=81
    PodCoreTissue=10259
    PodDeadTissue=10257
    PodMesoglea=10262
    PodOuterTissue=10260
    PodShellTissue=10261
    PodSurfaceTissue=10258
    PodTissue=10263
    PoliticalPrisoner=177
    Polymers=26
    PowerConverter=153
    PowerGenerators=83
    PowerTransferBus=161
    Praseodymium=143
    PreciousGems=10165
    ProgenitorCells=36
    ProhibitedResearchMaterials=10220
    PrototypeTech=130
    Pyrophyllite=112
    RadiationBaffle=162
    RareArtwork=131
    ReactiveArmour=80
    RebelTransmissions=132
    ReinforcedMountingPlate=163
    ResonatingSeparators=69
    Rhodplumsite=10243
    Robotics=70
    RockforthFertiliser=10264
    Rutile=59
    Samarium=142
    SAP8CoreContainer=90
    ScientificResearch=178
    ScientificSamples=179
    Scrap=77
    Semiconductors=28
    Serendibite=10244
    Silver=46
    SkimmerComponents=104
    Slaves=53
    SmallSurveyDataCache=10208
    SpacePioneerRelics=10164
    StructuralRegulators=117
    Superconductors=27
    SurfaceStabilisers=97
    SurvivalEquipment=164
    SyntheticFabrics=75
    SyntheticMeat=23
    SyntheticReagents=98
    Taaffeite=120
    TacticalData=180
    Tantalum=47
    Tea=22
    TechnicalBlueprints=133
    TelemetrySuite=184
    Thallium=108
    ThargoidBasiliskTissueSample=10236
    ThargoidBiologicalMatter=10160
    ThargoidCyclopsTissueSample=10234
    ThargoidHeart=10235
    ThargoidHydraTissueSample=10239
    ThargoidLink=10161
    ThargoidMedusaTissueSample=10237
    ThargoidProbe=186
    ThargoidResin=10162
    ThargoidScoutTissueSample=10238
    ThargoidSensor=10226
    ThargoidTechnologySamples=10163
    ThermalCoolingUnits=105
    Thorium=109
    TimeCapsule=10212
    Titanium=48
    Tobacco=13
    ToxicWaste=54
    TradeData=134
    TrinketsofHiddenFortune=135
    Tritium=10269
    UnstableDataCore=176
    Uraninite=60
    Uranium=50
    VoidOpals=10250
    Water=139
    WaterPurifiers=82
    Wine=14
    WreckageComponents=10207
is_normalized_key = lambda key:(not str(key).startswith('__')) and (not str(key).endswith('__'))
all_commodities = [k for k in Commodities.__dict__.keys() if (is_normalized_key(k))]
commodities_by_name = commodities = dict([tuple([k,v]) for k,v in Commodities.__dict__.items() if (is_normalized_key(k))])

commodities_by_value = {}
for k,v in commodities_by_name.items():
    if (v in commodities_by_value.keys()):
        commodities_by_value[v] = [commodities_by_value.get(v, k)]
    item = commodities_by_value.get(v, k)
    if (isinstance(item, list)):
        item.append(k)
    commodities_by_value[v] = item


django-standalone-0.4/example_script.py
#!/usr/bin/env python
# -*- coding: utf8 -*-

# parse command line to get the database and wether we want to
# create the database.
from optparse import OptionParser

parser = OptionParser("usage: %prog -d DATABASE [-s]")
parser.add_option('-d', '--database', dest='database', help="DATABASE file name", metavar="DATABASE")
parser.add_option('-s', '--syncdb', dest='syncdb', action="store_true", help="should the database be created?")
parser.add_option('-r', '--repl', dest='repl', action="store_true", help="start a REPL with access to your models")

options, args = parser.parse_args()

if not options.database:
    parser.error("You must specify the database name")

# fetch the settings and cache them for later use
from standalone.conf import settings
settings = settings(
    DATABASE_ENGINE='sqlite3',
    DATABASE_NAME=options.database,
)

# build the models we want to have in the database
from standalone import models

class MyModel(models.StandaloneModel):

    col1 = models.CharField(max_length=1000)
    col2 = models.IntegerField()
    col3 = models.BooleanField()

    def __unicode__(self):
        return self.col1

from django.core.management import call_command

if options.syncdb:
    # run a simple command - here syncdb - from the management suite
    call_command('syncdb')
elif options.repl:
    # start the shell, access to your models through import standalone.models
    call_command('shell')
else:
    l = MyModel.objects.filter(col1='foo')
    if len(l):
        print "found", l[0]
    else:
        o = MyModel(col1='foo', col2=5, col3=3)
        o.save()
        print "new", o



libs/inara.py
__copyright__ = """\
(c). Copyright 2008-2020, Vyper Logix Corp., All Rights Reserved.

Published under Creative Commons License 
(http://creativecommons.org/licenses/by-nc/3.0/) 
restricted to non-commercial educational use only., 

http://www.VyperLogix.com for details

THE AUTHOR VYPER LOGIX CORP DISCLAIMS ALL WARRANTIES WITH REGARD TO
THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND
FITNESS, IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL,
INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING
FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT,
NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
WITH THE USE OR PERFORMANCE OF THIS SOFTWARE !

USE AT YOUR OWN RISK.
"""
import os
import sys
import math

class CommodityError(Exception):
    pass

class CommodityValueError(Exception):
    pass

def sample():
    """
    $apitestdata = array();   // your data should be there
    $jsondata = json_encode($apitestdata);
    $url = "https://inara.cz/inapi/v1/";
    $ch = curl_init($url);
    curl_setopt($ch, CURLOPT_HTTPHEADER, array("Content-Type: application/json"));
    curl_setopt($ch, CURLOPT_POST, 1);
    curl_setopt($ch, CURLOPT_POSTFIELDS, $jsondata);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_TIMEOUT, 30);
    $response = curl_exec($ch);
    curl_close($ch);
    $result = json_decode($response, true);   // process the result as you wish
    """
    import requests
    import simplejson

    apidata = """
    {
        "header": {
            "appName": "TheCappySmackApp",
            "appVersion": "0.1",
            "isDeveloped": true,
            "APIkey": "7uhx0rc25zocwco0w8w48ockw0ogcswoscwo4gk",
            "commanderName": "CappySmack",
            "commanderFrontierID": "3478314"
        },
        "events": [
            {
                "eventName": "getCommanderProfile",
                "eventTimestamp": "2017-05-02T17:30:49Z",
                "eventData": {
                    "searchName": "CappySmack"
                }
            }
        ]
    }
"""
    apidata = ''.join([ll for ll in [str(l).strip() for l in apidata.split('\n')] if (len(ll) > 0)])
    payload_data =  simplejson.loads(apidata)
    payload =  simplejson.dumps(payload_data)
    url = "https://inara.cz/inapi/v1/"

    headers = {
        'user-agent': 'my-app/0.0.1',
        'Content-Type': 'application/json'
    }
    r = requests.post(url, data=payload, headers=headers)
    if (r.ok):
        response = r.json()
        print(response)

    #$result = json_decode($response, true);   // process the result as you wish


def scrape_commodities():
    import os
    import requests

    from vyperlogix import enum

    from bs4 import BeautifulSoup

    from io import StringIO

    commodities = {}

    __tritium__ = 'Tritium'

    commodities_url1 = "https://inara.cz/galaxy-commodities/"

    r = requests.get(commodities_url1)
    if (r.ok):
        soup = BeautifulSoup(r.content, 'html.parser')
        divs = soup.findAll(name='div', attrs={'class':"maincontentcontainer"})
        selects = soup.findAll(name='select', attrs={'name':'searchcommodity'})
        for sel in selects:
            for child in sel.children:
                if (child.name.lower() == 'option'):
                    if (child.getText() not in commodities.keys()):
                        commodities[child.getText()] = child.attrs.get('value', None)
                    print('{} --> {}'.format(child.getText(), child.attrs))
        if (__tritium__ not in commodities.keys()):
            raise(CommodityError, 'Cannot find {}.'.format(__tritium__))
        trit = commodities.get(__tritium__, None)
        if (not trit):
            raise(CommodityValueErrorError, 'Cannot find value for {}.'.format(__tritium__))
        fname = os.path.abspath('./libs/commodities.py')
        with open(fname, 'w') as ffOut:
            fOut = StringIO()
            fOut.write('from vyperlogix import enum\n')
            fOut.write('class Commodities(enum.Enum.Enum):\n')
            for k,v in commodities.items():
                fOut.write('{}{}={}\n'.format(' '*4,''.join(''.join(''.join(k.split()).split('-')).split('.')),v))
            fOut.write('')
            fOut.write('')
            fOut.write("is_normalized_key = lambda key:(not str(key).startswith('__')) and (not str(key).endswith('__'))\n")
            fOut.write("all_commodities = [k for k in Commodities.__dict__.keys() if (is_normalized_key(k))]\n")
            fOut.write("commodities = dict([tuple([k,v]) for k,v in Commodities.__dict__.items() if (is_normalized_key(k))])\n")
            fOut.write('')
            fOut.flush()
            ffOut.writelines(fOut.getvalue())
            fOut.close()
    else:
        print('WARNING: Problem with {} {}.'.format(commodities_url1, r.status_code))


def upload_to_google_drive(fname):
    import os
    from vyperlogix import _utils
    from pydrive.auth import GoogleAuth
    from pydrive.drive import GoogleDrive
    
    gauth = GoogleAuth()
    my_creds_fpath = os.path.abspath('./mycreds.txt')
    if (not os.path.exists(my_creds_fpath)) or (not os.path.isfile(my_creds_fpath)):
        gauth.LocalWebserverAuth() # bootstrap to achieve the desired goal.
        gauth.SaveCredentialsFile(my_creds_fpath)
    else:
        gauth.LoadCredentialsFile(my_creds_fpath)
    
    drive = GoogleDrive(gauth)    

    target_drive_folder = '@data'
    try:
        folders = drive.ListFile({'q': " title='{}' ".format(target_drive_folder)}).GetList()
    except Exception as ex:
        print('WARNING: {}.'.format(ex))
        folders = []

    for folder in folders:
        if folder['title'] == target_drive_folder:
            target_filename = '{}_{}'.format(os.path.basename(fname), _utils.timeStampForFileName())
            print('{} {} --> {}'.format(folder.get('id'), folder.get('title'), target_filename))
            file2 = drive.CreateFile({'parents': [{'id': folder['id']}], 'title': target_filename})
            file2.SetContentFile(fname)
            file2.Upload()    
    print('\n')
        
        
def scrape_commodity_data(commodity_refid=10269, star_system_refid=0, dirname='./data', is_verbose=False, fOut=sys.stdout):
    """
    fOut can be sys.stdout or a Queue to allow for background processing.
    """
    import requests

    from vyperlogix import _utils
    
    import commodities

    from bs4 import BeautifulSoup

    __location__ = 'Location'

    __special_cols__ = [__location__, 'Buy price', 'Sell price', 'QTY', 'St dist', 'Distance']
    
    normalize_key = lambda k:'{}{}'.format('_'.join([t for t in str(k).split() if (len(t) > 0)]),'_value')
    
    __commodity_name__ = commodities.commodities_by_value.get(commodity_refid, None)
    if (__commodity_name__ is None) or (not isinstance(__commodity_name__, str)):
        raise(CommodityValueError, 'WARNING: Cannot resolve {} to a valid Commodity Name.'.format(commodity_refid))
    
    #__tritium__ = 'Tritium'

    #tritium_url = "https://inara.cz/ajaxaction.php?act=goodsdata&refname=buymin&refid=10269&refid2=367"     # refid is the commodity, refid2 is the star system.
    #tritium_url2 = "https://inara.cz/ajaxaction.php?act=goodsdata&refname=buymin&refid=10269&refid2=18698"
    #tritium_url3 = "https://inara.cz/ajaxaction.php?act=goodsdata&refname=sellmax&refid=10269&refid2=18698"

    #commodities_url1 = "https://inara.cz/galaxy-commodity/10269/"
    commodities_buymin_url = "https://inara.cz/ajaxaction.php?act=goodsdata&refname=buymin&refid={}&refid2={}".format(commodity_refid, star_system_refid)
    commodities_sellmax_url = "https://inara.cz/ajaxaction.php?act=goodsdata&refname=sellmax&refid={}&refid2={}".format(commodity_refid, star_system_refid)


    def format_header(row, use_keys=True):
        formatted_header = []
        try:
            subj = row.keys() if (use_keys) else row.values()
            fmt = ['{}'.format(item) for item in subj]
            for item in fmt:
                formatted_header.append(item)
            formatted_header = '\t'.join(formatted_header).strip()+'\n'
        except Exception as ex:
            formatted_header = '-- INVALID --'
        return formatted_header


    def print_data(the_data, verbose=False):
        from io import StringIO

        fOut = StringIO()
        fOut.write('BEGIN:\n')
        has_header = False
        try:
            for row in the_data:
                if (len(row.keys()) > 0):
                    if (not has_header):
                        fOut.write(format_header(row, use_keys=True))
                        has_header = True
                    fOut.write(format_header(row, use_keys=False))
        except Exception as ex:
            print(ex)
        fOut.flush()
        lines = fOut.getvalue().split('\n')
        fOut.write('END!!!\n')
        for l in lines:
            if (verbose):
                print(l)
        fOut.close()
        
        
    def export_as_csv_file(the_headers, the_data, dirname, basefilename, verbose=False):
        import os
        import csv
        try:
            fpath = os.path.abspath(dirname)
            if (not os.path.isdir(fpath)):
                os.mkdir(fpath)
            fname = os.path.abspath(os.path.join(fpath, basefilename))
            with open(fname,mode='w',encoding='utf8',newline='') as output_to_csv:
                dict_csv_writer = csv.DictWriter(output_to_csv, fieldnames=the_headers,dialect='excel')
                dict_csv_writer.writeheader()
                dict_csv_writer.writerows(the_data)
            if (verbose):
                print('\nData exported to {} succesfully and sample data'.format(fname))
        except IOError as io:
            print('\n',io)
        
        
    def harmonic_averages(dataFrame, col_name, harmonic=2, qty_col_name=None, min_qty=1000, lower_bound=None, upper_bound=None):
        """
        This could be used to produce a set of bands of selectioned regions to help determine regions of interest.
        """
        subDataFrame = dataFrame
        new_lower_bound = dataFrame[col_name].min()
        new_upper_bound = dataFrame[col_name].max()
        if (lower_bound):
            subDataFrame = subDataFrame[col_name] >= lower_bound
        if (upper_bound):
            subDataFrame = subDataFrame[col_name] <= upper_bound
        if (qty_col_name):
            subDataFrame = subDataFrame[(subDataFrame[qty_col_name] >= min_qty)]
        while (True):
            count = subDataFrame[col_name].count()
            sample_mean = subDataFrame[col_name].mean()
            subDataFrame = subDataFrame[subDataFrame[col_name] < sample_mean]
            new_count = subDataFrame[col_name].count()
            if (new_count < (count / harmonic)):
                new_lower_bound = subDataFrame[col_name].min()
                new_upper_bound = subDataFrame[col_name].max()
                break
        return dataFrame[(dataFrame[col_name] >= new_lower_bound) & (dataFrame[col_name] <= new_upper_bound)]


    def fetch_data_from(url, dirname=None, filename=None, is_verbose=False, is_debugging=False, is_uploading=False, filter_keys_for_callback=[], callback=None):
        the_headers = set()
        the_results = []
        r = requests.get(url)
        if (r.ok):
            soup = BeautifulSoup(r.content, 'html.parser')
            tables = soup.findAll(name='table', attrs={'class':"tablesorter"})
            for table in tables:
                headers = [header.text for header in table.find_all('th')]
                results = [{headers[i]: cell for i, cell in enumerate(row.find_all('td'))}
                           for row in table.find_all('tr')]
                for result in results:
                    if (len(result.keys()) > 0):
                        row = {}
                        for k,v in result.items():
                            items_in_row = []
                            try:
                                for item in [c for c in v.children]:
                                    try:
                                        items_in_row.append(item.text)
                                    except Exception as ex:
                                        items_in_row.append(str(item))
                                row[k] = ''.join(items_in_row)
                            except Exception as ex:
                                print(ex)
                        the_results.append(row)
                break
    
            # filter the buy price to ensure the value is float
            new_results = []
            for row in the_results:
                use_the_row = True
                for k in filter_keys_for_callback:
                    fuzzy_row_keys = dict([tuple([kk.lower(),kk]) for kk in row.keys()])
                    if (k is None):
                        continue
                    kk = fuzzy_row_keys.get(k.lower(), None)
                    if (kk):
                        k = kk
                        did_callback_happen = False
                        if (not _utils.is_floating_or_numeric_digits(row.get(k, 0.0))):
                            new_key = normalize_key(k)
                            try:
                                if (callable is not None) and (callable(callback)):
                                    new_key, new_value = callback(k, row.get(k), filter_keys_for_callback)
                                    did_callback_happen = True
                            except Exception as ex:
                                print(ex)
                        if (not did_callback_happen):
                            new_value = row.get(k, 0.0)
                            if (not isinstance(new_value, float)):
                                try:
                                    new_value = ''.join([c for c in new_value if (_utils.is_floating_or_numeric_digit(c))]).replace('...', '').replace('..', '')
                                    new_value = float(new_value) if (_utils.is_floating_or_numeric_digits(new_value)) else 0.0
                                except Exception as ex:
                                    sys.stderr.write('{}\n'.format(ex))
                        if ((new_key is not None) and (new_value is not None)):
                            row[new_key] = new_value
                            if (is_debugging):
                                print('{} -> {} --> {} -> {}'.format(k, row.get(k, 0.0), new_key, row.get(new_key, 0.0)))
                        else:
                            use_the_row = False
                            break
                if (use_the_row):
                    new_results.append(row)
    
            the_results = new_results
            if (is_verbose):
                print_data(the_results, verbose=is_verbose)        
    
            if (dirname is not None) and (filename is not None):
                for item in the_results:
                    the_headers = the_headers.union(set(item.keys()))
                export_as_csv_file(list(the_headers), the_results, dirname, filename, verbose=is_verbose)
        
                if (is_uploading):
                    upload_to_google_drive(fname) # must debug this.
        else:
            print('WARNING: Problem with {} {}.'.format(url, r.status_code))

        return the_headers, the_results
    
    
    def special_column_filter(k, v, special_cols):
        new_key, new_value = _utils.ascii_only(k) if (isinstance(k, str)) else k, _utils.ascii_only(v) if (isinstance(v, str)) else v
        if (isinstance(k, str)) and (isinstance(v, str)) and (k.lower() == str(__location__).lower()):
            import re
            regex = r"(\([A-Za-z0-9]{3}-[A-Za-z0-9]{3}\))"
            matches = [m for m in re.finditer(regex, v, re.MULTILINE)]
            if (len(matches) > 0):
                new_key = new_value = None # signal the skipping of this row - skipped over Carriers.
        else:
            new_key = normalize_key(k)
            new_value = ''.join([c for c in v if (_utils.is_floating_or_numeric_digit(c))]) if (isinstance(v, str)) else v
            new_value = float(new_value) if (isinstance(new_key, float)) or (_utils.is_floating_or_numeric_digits(new_value)) else 0.0
            
        if (new_key is not None):
            new_special_cols = list(set(special_cols).union(set([new_key])))
            while (len(special_cols) > 0):
                special_cols.pop()
            for item in new_special_cols:
                special_cols.append(item)
        return tuple([new_key, new_value])


    def write_output(fOut, msg):
        try:
            if (isinstance(msg, str)):
                fOut.write(msg + '\n')
            else:
                fOut.write(msg.to_string() + '\n')
        except:
            try:
                fOut.put(msg if (isinstance(msg, str)) else msg.ro_string())
            except Exception as ex:
                sys.stderr.write(msg + '\n')


    def normalize_frame_keys(frame, specials):
        sCols = []
        try:
            keys = frame.keys()
            for col in specials:
                if (col in keys):
                    sCols.append(col)
        except Exception as ex:
            sys.stderr.write('{}\n'.format(ex))
        return sCols
    
    import pandas

    pandas.set_option('display.max_columns', None)
    pandas.set_option('display.width', 200)

    the_headers, buymin_results = fetch_data_from(commodities_buymin_url, dirname=dirname, filename='commodity_{}_buymin.csv'.format(__commodity_name__), is_verbose=False, is_debugging=False, filter_keys_for_callback=__special_cols__, callback=special_column_filter)
    
    df_buymin = pandas.DataFrame(buymin_results)

    colName = 'Buy_price_value'
    #print('(0) Min {}, Max {}, Mean {}'.format(df_buymin[colName].min(), df_buymin[colName].max(), df_buymin[colName].mean()))

    buymin_subDataFrame = harmonic_averages(df_buymin, colName, harmonic=3, min_qty=1000, qty_col_name='QTY_value')
    if (is_verbose):
        write_output(fOut, '(1) Min {}, Max {}, Mean {}'.format(buymin_subDataFrame[colName].min(), buymin_subDataFrame[colName].max(), buymin_subDataFrame[colName].mean()))
    buymin_price = buymin_subDataFrame[colName].min()
    if (is_verbose):
        write_output(fOut, buymin_subDataFrame[normalize_frame_keys(buymin_subDataFrame, __special_cols__)].dropna())

    fpath = os.sep.join([dirname, 'commodity_{}_buymin_locations.csv'.format(__commodity_name__)])
    with open(fpath, 'w') as ffOut:
        buymin_subDataFrame[normalize_frame_keys(buymin_subDataFrame, __special_cols__)].to_csv(sys.stdout if (is_verbose) else ffOut)
        ffOut.flush()

    the_headers, sellmax_results = fetch_data_from(commodities_sellmax_url, dirname=dirname, filename='commodity_{}_sellmax.csv'.format(__commodity_name__), is_verbose=False, is_debugging=False, filter_keys_for_callback=__special_cols__, callback=special_column_filter)
    
    df_sellmax = pandas.DataFrame(sellmax_results)

    colName = 'Sell_price_value'
    #print('(0) Min {}, Max {}, Mean {}'.format(df_sellmax[colName].min(), df_sellmax[colName].max(), df_sellmax[colName].mean()))

    sellmax_subDataFrame = harmonic_averages(df_sellmax, colName, harmonic=3, min_qty=1000, qty_col_name='QTY_value')
    if (is_verbose):
        write_output(fOut, '(1) Min {}, Max {}, Mean {}'.format(sellmax_subDataFrame[colName].min(), sellmax_subDataFrame[colName].max(), sellmax_subDataFrame[colName].mean()))
    sellmax_price = sellmax_subDataFrame[colName].min()
    
    try:
        fpath = os.sep.join([dirname, 'commodity_{}_sellmax_locations.csv'.format(__commodity_name__)])
        isnan_sellmax_price = math.isnan(sellmax_price)
        isnan_buymin_price = math.isnan(buymin_price)
        if (not isnan_sellmax_price) and (not isnan_buymin_price) and (int(sellmax_price / buymin_price) > 5):
            if (is_verbose):
                write_output(fOut, 'Locations with profit margin greater than 5:')
                write_output(fOut, sellmax_subDataFrame[normalize_frame_keys(sellmax_subDataFrame, __special_cols__)].dropna())
            with open(fpath, 'w') as ffOut:
                sellmax_subDataFrame[normalize_frame_keys(sellmax_subDataFrame, __special_cols__)].to_csv(sys.stdout if (is_verbose) else ffOut)
                ffOut.flush()
        elif (isnan_sellmax_price) or (isnan_buymin_price):
            with open(fpath, 'w') as ffOut:
                ffOut.write('WARNING: {}{}{}\n'.format('Max sell price is NaN.' if (isnan_sellmax_price) else '', ' or ' if (isnan_sellmax_price or isnan_buymin_price) else ' and ' if (isnan_sellmax_price and isnan_buymin_price) else '', 'Min buy price is NaN.' if (isnan_buymin_price) else ''))
                ffOut.flush()
    except Exception as ex:
        sys.stderr.write('WARNING: {}'.format(ex))

    #print(sellmax_subDataFrame)          
        
if (__name__ == '__main__'):
    scrape_commodity_data()
